File: javascript/resistor-color-trio/resistor-color-trio.js
//
// This is only a SKELETON file for the 'Resistor Color Trio' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

const colors = {
  'black': 0,
  'brown': 1,
  'red': 2,
  'orange': 3,
  'yellow': 4,
  'green': 5,
  'blue': 6,
  'violet': 7,
  'grey': 8,
  'gray': 8,
  'white': 9,
}

export class ResistorColorTrio {
  constructor(bands) {
    const values = bands.map((band) => colors[band])
    if (values.length !== bands.length || values.some((value) => typeof value === 'undefined'))
      throw new Error('invalid color')

    let amount = 0
    for(let i = 0; i < values.length - 1; i++)
      if (i >= 2)
        amount *= 10
      else 
        amount = (amount * 10) + values[i]

    let  multiplier = values[values.length - 1]
    amount *= Math.pow(10, multiplier)
    let unit = 'ohms'

    if (amount >= 1_000_000_000)
    {
      unit = 'gigaohms'
      amount /= 1_000_000_000
    }
    else if (amount >= 1_000_000)
    {
      unit = 'megaohms'
      amount /= 1_000_000
    }
    else if (amount >= 1_000)
    {
      unit = 'kiloohms'
      amount /= 1_000
    }

    this._label = `Resistor value: ${amount} ${unit}`

  }

  get label() {
    return this._label
  }
}

File: go/resistor-color-trio/resistor_color_trio.go
// For a trio of resistor colors calculate the amount and prefix and return the value as a human readable string.
package resistorcolortrio

import (
	"fmt"
	"math"
)

// Resistor color to band value.
var colorLookup map[string]int = map[string]int{
	"black":  0,
	"brown":  1,
	"red":    2,
	"orange": 3,
	"yellow": 4,
	"green":  5,
	"blue":   6,
	"violet": 7,
	"grey":   8,
	"white":  9,
}

// index = how many times we add an extra three zeros which gives the proper prefix.
var prefix map[int]string = map[int]string{
	0: "",
	1: "kilo",
	2: "mega",
	3: "giga",
}

// Label describes the resistance value given the colors of a resistor.
// The label is a string with a resistance value with an unit appended
// (e.g. "33 ohms", "470 kiloohms").
// @param colors: slice of color bands (only first three will be used)
// @returns: String representation of resistance with a number then ohm rating in (/kilo/mega/giga)ohms
func Label(colors []string) string {
	amount := 0
	for i := 0; i < min(2, len(colors)); i++ {
		amount = (amount * 10) + colorLookup[colors[i]]
	}
	if len(colors) >= 3 {
		amount *= int(math.Pow(10.0, float64(colorLookup[colors[2]])))
	}
	prefixIdx := 0
	for prefixIdx < len(prefix) && amount > 1000 {
		prefixIdx++
		amount /= 1000
	}

	return fmt.Sprintf("%d %sohms", amount, prefix[prefixIdx])
}
